node/src/components/block_validator.rs

```diff
@@ -186,33 +241,26 @@ where
                 block,
                 sender,
                 responder,
-                block_timestamp,
             }) => {
-                let block_deploys = block.deploys();
-                let deploy_count = block_deploys.len();
-                // Collect the deploys in a set; this also deduplicates them.
-                let block_deploys: HashSet<_> = block_deploys
-                    .iter()
-                    .map(|deploy_hash| **deploy_hash)
-                    .collect();
+                let deploy_count = block.deploy_hashes().len() + block.transfer_hashes().len();
+                if deploy_count == 0 {
+                    // If there are no deploys, return early.
+                    return responder.respond(true).ignore();
+                }
+                // Collect the deploys in a set. If they are fewer now, then there was a duplicate!
+                let block_deploys: HashSet<_> = block.deploys_and_transfers_iter().collect();
                 if block_deploys.len() != deploy_count {
                     self.log_block_with_replay(sender, &block);
-                    return responder.respond((false, block)).ignore();
-                }
-                if block_deploys.is_empty() {
-                    // If there are no deploys, return early.
-                    return responder.respond((true, block)).ignore();
+                    return responder.respond(false).ignore();
                 }
 
-                // TODO: Clean this up to use `or_insert_with_key` once
-                // https://github.com/rust-lang/rust/issues/71024 is stabilized.
                 match self.validation_states.entry(block) {
                     Entry::Occupied(mut entry) => {
                         // The entry already exists.
```

Comment:
Note: this may be unreachable, but this is the correct behavior even in the case that we do reach it.